k-distribution/k-tutorial/1_basic/22_proofs/README.md

```diff
@@ -1 +1,444 @@
+
+2. The second claim reasons about the `if` statement that has a concrete condition:
+
+```
+claim <k> if ( 3 + 4 == 7 ) {
+            $a = 1 ;
+            } else {
+            $a = 2 ;
+            }
+        => . ... </k>
+        <store> STORE => STORE [ $a <- 1 ] </store>
+```
+
+stating that the given program terminates (`=> .`), and when it does, the value
+of the variable `$a` is set to `1`, meaning that the execution will have taken
+the `then` branch. Add this claim to the `LESSON-22-SPEC` module, but also add
+
+```
+syntax Id ::= "$a" [token]
+```
+
+to the `LESSON-22-SPEC-SYNTAX` module in order to declare `$a` as a token so
+that it can be used as a program variable. Re-run the K prover, which should
+again return `#Top`.
+
+3. Our third claim demonstrates how to reason about both branches of an `if`
+   statement at the same time:
+
+```
+claim <k> $a = A:Int ; $b = B:Int ;
+          if (A < B) {
+            $c = $b ;
+          } else {
+            $c = $a ;
+          }
+        => . ... </k>
+        <store> STORE => STORE [ $a <- A ] [ $b <- B ] [ $c <- ?C:Int ] </store>
+    ensures (?C ==Int A) orBool (?C ==Int B)
+```
+
+The program in question first assigns symbolic integers `A` and `B` to program
+variables `$a` and `$b`, respectively, and then executes the given `if`
+statement, which has a symbolic condition (`A < B`), updating the value of the
+program variable `$c` in both branches. The specification we give states that
+the `if` statement terminates, with `$a` and `$b` updated, respectively, to `A`
+and `B`, and `$c` updated to *some* symbolic integer value `?C`. Via the
+`ensures` clause, which is used to specify additional constraints that hold
+after execution, we also state that this existentially quantified `?C` equals
+either `A` or `B`.
+
+Add the productions declaring `$b` and `$c` as tokens to the
+`LESSON-22-SPEC-SYNTAX` module, the claim to the `LESSON-22-SPEC` module, run
+the K prover again, and observe the output, which should not be `#Top` this
+time. This means that K was not able to prove the claim, and we now need to
+understand why. We do so by examining the output, which should look as follows:
+
+```
+    (InfoReachability) while checking the implication:
+    The configuration's term unifies with the destination's term,
+    but the implication check between the conditions has failed.
+
+  #Not (
+    #Exists ?C . {
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- ?C:Int ]
+      #Equals
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    }
+  #And
+    {
+      true
+    #Equals
+      ?C ==Int A orBool ?C ==Int B
+    }
+  )
+#And
+  <generatedTop>
+    <k>
+      _DotVar1
+    </k>
+    <store>
+      STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    </store>
+    <funcs>
+      _Gen3
+    </funcs>
+    <stack>
+      _Gen5
+    </stack>
+  </generatedTop>
+#And
+  {
+    true
+  #Equals
+    A <Int B
+  }
+```
```

Comment:
On Friday when I tested this I was only getting this output after including `MAP-SYMBOLIC`, otherwise it was getting stuck earlier on some reasoning about maps. Maybe this is just caused by a quirk in my setup.

Reply:
Ahhh, it might be that @charala1 removing `domains.md` from being imported [here](https://github.com/runtimeverification/k/pull/2699/commits/9628cfa4b69ed317a5908014c9eb537fa8b114ee) caused this. I think we need that back.